pkg/conduit/pipelines_parser.go

```diff
@@ -0,0 +1,75 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package conduit wires up everything under the hood of a Conduit instance
+// including metrics, telemetry, logging, and server construction.
+// It should only ever interact with the Orchestrator, never individual
+// services. All of that responsibility should be left to the Orchestrator.
+package conduit
+
+import (
+	"bytes"
+	"fmt"
+	"gopkg.in/yaml.v3"
+)
+
+type Pipelines []struct {
+	Status string `yaml:"status"`
```

Comment:
I think we will need an identifier field for each entity (pipeline, connector, processor) to be able to distinguish them between runs. For example, if a config file changes and we reload it we need to be able to know which entity changed, because we will have the previous pipeline config already stored somewhere. Or if a config file is removed we need to know which pipeline to remove. All of that can't be deterministically done without an identifier field (happy to hear suggestions if you think it is 🙂).

I see two options:
- either we add an `id` field to each entity
  ```
  connectors:
    - id: my-source
      type: source
      ...
    - id: my-destination
      type: destination
      ...
  ```
- or we change the array into a map, where the key of the map is the ID of the connector/processor
  ```
  connectors:
    my-source:
      type: source
      ...
    my-destination:
      type: destination
      ...
  ```

The second one looks nicer to me. If we decide on the second option, I think it might make sense to add a new root element `pipelines` and allow creating multiple in the same file (I was against that at first, but with this approach it seems it would make for a more complete design). For example:
```
pipelines:
  my-pipeline:
    state: running
    connectors:
      my-source:
        type: source
        ...
      my-destination:
        type: destination
        ...
    processors:
      my-processor:
        ...
  my-other-pipeline:
    ...
```

WDYT @ConduitIO/conduit-core?

Reply:
IMHO, that depends on the use case. Conduit always being in sync with the files appears to be the problem which we want to solve with the above. For v1 I think it's useful to just have the ability to load pipeline(s) and either fail or overwrite if a pipeline with the same config exists.

Reply:
Also, re: status: the only two status which we can have here are actually "stopped" and "running". Having a degraded pipeline in a config is a bit strange.

Reply:
> fail or overwrite if a pipeline with the same config exists

The "overwrite" part is what I'm talking about. If we just overwrite (i.e. delete the previous pipeline and create a new one) we lose the state of the pipeline. I'm not talking about periodically hot reloading files (although that would also be nice), but restarting Conduit and reloading pipeline config files - we still need to be sure that we don't lose the position the pipeline processed last.

> Having a degraded pipeline in a config is a bit strange.

I was thinking about this as well, almost wrote a comment about it but didn't after all. First of all, "stopped" isn't the same as "degraded", because there would be no error stored in the pipeline. I think it might be useful for cases when you want to temporarily pause a pipeline but keep it created in Conduit for some later time. Or maybe it's a pipeline that you want to have prepared and start/stop manually on demand. So I see some use cases for this field.

That said, I think the field should be optional and should default to "running" if not specified otherwise (that shouldn't be part of the parser).

Reply:
I think having an id for each entity is good in general, especially because we need the id for the parent in the processors and we didn't have that, also for the reason lovro mentioned (even if we dismiss it for the first version).
 About the preferred format for the yaml file, I don't have a strong preference honestly, first one looks easier for the user to fill {it's like a form that you have all the questions (the keys) and you fill the value}, the second approach is more of you decide what the name of the key is, and fill the value, which would only need from us to have good documentation showing that the id has to be unique.